pkg/go/graph/weighted_graph.go

```diff
@@ -744,3 +798,285 @@ func (wg *WeightedAuthorizationModelGraph) removeNodeFromTupleCycles(nodeID stri
+			// if the node weight is 0 then the userset is not found in this path, otherwise the weight will be calculated appropiately
+			if nodeWeight == 0 {
+				wg.setUsersetWeightToEdge(edge, usersetNode.uniqueLabel, 0)
+			} else {
+				if nodeWeight != Infinite && (edge.edgeType == DirectEdge || edge.edgeType == TTUEdge) {
+					wg.setUsersetWeightToEdge(edge, usersetNode.uniqueLabel, nodeWeight+1)
+				} else {
+					wg.setUsersetWeightToEdge(edge, usersetNode.uniqueLabel, nodeWeight)
+				}
+				foundWeight = true
+			}
+		}
+	}
+
+	weight := 0
+	if foundWeight {
+		weight = Infinite
+	}
+
+	wg.setUsersetWeightToNode(node, usersetNode.uniqueLabel, weight)
+	return weight
+}
+
+func (wg *WeightedAuthorizationModelGraph) calculateUsersetWeightMaxStrategy(node *WeightedAuthorizationModelNode, usersetNode *WeightedAuthorizationModelNode) int {
+	edges, ok := wg.GetEdgesFromNode(node)
+	if !ok {
+		return wg.setUsersetWeightToNode(node, usersetNode.uniqueLabel, 0)
+	}
+
+	maxWeight := 0
+	for _, edge := range edges {
+		if edge.usersetWeights[usersetNode.uniqueLabel] > 0 && edge.usersetWeights[usersetNode.uniqueLabel] > maxWeight {
+			maxWeight = edge.usersetWeights[usersetNode.uniqueLabel]
+		}
+	}
+	return wg.setUsersetWeightToNode(node, usersetNode.uniqueLabel, maxWeight)
+}
+
+func (wg *WeightedAuthorizationModelGraph) calculateUsersetWeightMaxStrategyWithEnforcement(node *WeightedAuthorizationModelNode, usersetNode *WeightedAuthorizationModelNode) int {
+	edges, ok := wg.GetEdgesFromNode(node)
+	if !ok {
+		return wg.setUsersetWeightToNode(node, usersetNode.uniqueLabel, 0)
+	}
+
+	maxWeight := 0
+	for _, edge := range edges {
+		if edge.usersetWeights[usersetNode.uniqueLabel] == 0 {
+			return wg.setUsersetWeightToNode(node, usersetNode.uniqueLabel, 0)
+		}
+		if edge.usersetWeights[usersetNode.uniqueLabel] > maxWeight {
+			maxWeight = edge.usersetWeights[usersetNode.uniqueLabel]
+		}
+	}
+	return wg.setUsersetWeightToNode(node, usersetNode.uniqueLabel, maxWeight)
+}
+
+func (wg *WeightedAuthorizationModelGraph) calculateUsersetWeightHybridMaxstrategy(node *WeightedAuthorizationModelNode, usersetNode *WeightedAuthorizationModelNode) int {
+	edges, ok := wg.GetEdgesFromNode(node)
+	if !ok || len(edges) != 2 || edges[0].usersetWeights[usersetNode.uniqueLabel] == 0 {
+		return wg.setUsersetWeightToNode(node, usersetNode.uniqueLabel, 0)
+	}
+
+	weight1 := edges[0].usersetWeights[usersetNode.uniqueLabel]
+	weight2 := edges[1].usersetWeights[usersetNode.uniqueLabel]
+	if weight2 == 0 || weight1 > weight2 {
+		return wg.setUsersetWeightToNode(node, usersetNode.uniqueLabel, weight1)
+	}
+
+	return wg.setUsersetWeightToNode(node, usersetNode.uniqueLabel, weight2)
+}
+
+// What edges we can prune when we don't know what edges will lead to the userset if any
+// 1- If the edge does not have a weight to any of the userset node's weights, we can prune it and set weight to the userset to be 0
```

Comment:
Since we don't do setting weight to the userset to be 0 in this function, let's not add this as comment. The responsibility of this function is to return whether the edge can be prune.  It is up to the caller to determine the action.